docs/madr/decisions/020-kds-redesign.md

```diff
@@ -0,0 +1,238 @@
+# KDS Redesign
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4354
+
+## Context and Problem Statement
+
+The current implementation of KDS is quite complex, and there may be some bugs, such as the one found at https://github.com/kumahq/kuma/pull/5373, which was very difficult to identify and resolve. The use of a bi-directional stream for configuration exchange between the zone and the global can also be challenging to implement. Additionally, KDS operates as a State of The World, meaning that whenever an item in the resource type changes, the entire state is sent. While this doesn't appear to be a problem when synchronizing data from Global to Zone, it may be less efficient when there are many Dataplane changes. There are several improvements we want to make in the future such as:
+* working with different older versions of the control plane on the zone,
+* creating resources on the zone and syncing them to the global,
+* creating separate gRPC services for synchronization between the zone and global,
+* supporting multiple control plane tenants,
+* and sending only changes instead of the state of the world when synchronizing data from the zone to the global.
```

Comment:
Are we also missing a point about separating `informational` and `operational` streams. For example I'd expect policies and ZoneIngress/ZoneEgress to be synced as quickly as possible. On the other hand for Insights and Dataplanes things can be done a slower pace (but more efficient)

Reply:
I forgot to write it down. I will update the docs. My idea to do this is by separating generation of resources in the code. I think maybe not as a first step but in the second iteration. Each request has information about the type of resource for which it requires an update. We can define a parameter that defines the time after insights are going to be regenerated, and each regeneration of insight is going to put them into the cache. In case of a version change, that would cause the update to be sent.

Reply:
But that would slow down generated insights on the zone which might not be desirable.
Also we may want to introduce `linger` to maximize batching of changes no? This may be tricky if we don't separate the 2 types

Reply:
If I understood correctly the code resources that are synchronized with KDS are using separate caches and should be [independent](https://github.com/kumahq/kuma/blob/master/pkg/kds/server/components.go#L98) of the one that the zone has generated.  I thought about stream per resource as we have now, do you mean using 2 streams of type `PriorityResources` and `InsightResources` that can be batched? For sure we can aggregate data better this way.

Reply:
Yeah that was my thinking. I guess names would need to be better. I'm unsure if it required to separate them but I feel like being able to improve KDS batching of resource updates when syncing dataplanes and insights would improve performance quite a bit.
I'm curious to hear what @jakubdyszkiewicz thinks

Reply:
I don't think we need to separate them on the gRPC level.
It's just a matter of how often we put resources in a SnapshotCache (or whatever go-control-plane's Cache)